### DISCUSSION TITLE
Perses Market Place

### DISCUSSION THREAD

**Discussion by nicolastakashi** [perses/perses#2439] [Ideas]
# Perses Marketplace Proposal

## Introduction

We have been discussing the concept of a **Perses Marketplace**, a platform where users can share and discover **Dashboards**, **Panels**, and other plugins such as **DataSources** and **Widgets** within the Perses community. 

The primary goal of this marketplace is to foster collaboration and empower the Perses ecosystem by providing a seamless way for users to share and utilize resources.

---

## Vision

The **Perses Marketplace** aims to:
- Provide an easy integration for the Perses community to share and discover resources.
- Enhance collaboration and innovation by making it easier for users to contribute and utilize shared resources.
- Establish a simple yet powerful platform that serves as a hub for all things related to Perses.

---

## Key Features

The proposed marketplace will include the following features:

1. **Resource Discovery**
   - Enable users to search, filter, and explore various Perses resources.
   - Streamline the process of finding dashboards, panels, data sources, and widgets.

2. **User Profiles**
   - Allow users to create profiles for registering on the marketplace.
   - Enable resource publishing and profile customization.

3. **Resource Management**
   - Provide tools for users to publish and manage their shared resources.
   - Allow contributors to update and maintain their offerings within the marketplace.

---

## Mockups

We have created mockups to showcase the potential design and functionality of the marketplace:

1. **[Home Page](https://v0.dev/chat/9NUMFow6RNz)**: The central hub for exploring and discovering resources.
2. **[Sign In and Sign Up](https://v0.dev/chat/zYUw3vs5G5W)**: User authentication and onboarding pages.
3. **[User Profile](https://v0.dev/chat/L8vGDLj2uNn)**: Personalized pages for users to manage their profiles and resources.
4. **[Resource Publish](https://v0.dev/chat/AM9ZT2P4vfu)**: Interface for publishing and managing resources.
5. **[Explore Page](https://v0.dev/chat/yPi1CTJkVdW)**: A page for searching and exploring available resources.

---

## Collaboration

@bripkens and I are particularly interested in this idea and are eager to collaborate to bring it to life. We welcome contributions and discussions to make the **Perses Marketplace** a reality.

**Comment by ibakshay** [comment]
This is awesome! 
I have gone through the mock-ups, and I'm particularly impressed with the User Profile page - it looks outstanding! 🚀 
However, I'm unable to open the Explore Page mockup as it's returning the below error. 

![image](https://github.com/user-attachments/assets/17c395a1-2485-447e-85db-284bd844e374)

**Reply by nicolastakashi** [reply]
Thanks @ibakshay 
It's fixed now 🙇🏽‍♂️

**Comment by bripkens** [comment]
Just braindumping some thoughts:

# Storage of resources

As a first step, we could focus on storing and presenting dashboards. For this, we don't need any form of 
user management. I did something similar in the past for a chaos-engineering solution:
 - [A website](https://hub.steadybit.com/action/com.steadybit.extension_prometheus.instance.metrics) exposes the available resources (in their case actions, targets, extensions…)
 - The resources themselves are maintained [within a GitHub repo](https://github.com/steadybit/reliability-hub-db/tree/main/actions/com.steadybit.extension_prometheus.instance.metrics).
 - The website itself (closed source, in that case) is just treating the repo as a database.

This approach has a few advantages:

 - There is no need for user management or a separate storage mechanism for dashboards, thus dramatically reducing the scope.
 - Contribution mechanism and licensing of resources are pretty standard.
 - Ownership of the top X dashboards is clearer. If not for a repo, it is up to some central person to update the most common dashboards, e.g., NodeExporterFull (or we would need to establish a notion of teams).

It does open up some questions, however:

 - How can we infer usage and popularity? We could do this based on views and download counts.

# Storage Format
What kind of dashboard JSON/YAML format do we persist? The Perses API one slightly differs from the Perses operator one. The operator one is slightly more recognizable (due to the `Perses…` prefix in the `kind` field. In the end, it might just be that the app exposes a capability to download in either format :) 

# Technical Approach
For [OTelBin](https://github.com/dash0hq/otelbin/tree/main/packages/otelbin), we used a plain Next.js / Vercel / Clerk app. This worked well, and considering that it comes with OOTB support for serverless functions, it is also really flexible. At this point, they even support Go-based serverless functions.
This marketplace would likely be within the free tiers of these services – but we at Dash0 would also happily cover the fees if any arise.
I would recommend this approach again any time :).

**Reply by nicolastakashi** [reply]
I really like the idea of using GitHub as a file storage solution. Even if we don’t fully adopt it for this purpose, I believe we should advocate for a maturity model for published resources. For example, users who define their resources as code and make them publicly available could earn a higher score on this maturity scale.

In the future, I envision providing a GitHub Action that enables users to publish their resources to the Perses Marketplace seamlessly. This could be achieved through a CLI tool, likely named percli, embedded within the GitHub Action.

Regarding tracking stats, we could enhance the website by adding features such as a download button, tracking download counts, page views, and perhaps implementing a rating or “stars” system for resources.

However, it’s worth noting that we won’t have control over users cloning the repository, which might limit some tracking capabilities.

I’m also considering whether using an S3-like service for storage would be significantly more complex compared to relying on GitHub.

**Reply by Nexucis** [reply]
> In the future, I envision providing a GitHub Action that enables users to publish their resources to the Perses Marketplace seamlessly. This could be achieved through a CLI tool, likely named percli, embedded within the GitHub Action.

With @AntoineThebaud we would like to provide a github action to wrap `percli` (at least for the dasbhoard as code). But there is no reason to enhance it to publish a plugin to a marketplace.

## Technical aspect

On my side, I would rather prefer to have something close to what we have in Perses in term of technology. For example, If we have a backend in Golang we could try to reuse the auth/autorisation system implemented in Perses. Like that we would support quite faster OAuth2 / OIDC.

I believe that even for resources stored into a GitHub (or in GitLab, because why not) repository, you will need to publish a light resource in the Market Place to make the link to that specific repository. And to be able to create this light resource, then a full user/permissions system will be required.

I prefer to use a file database rather than using a git repository as a database.

Regarding the dashboards, I think yeah it makes sense they are stored in an archive in a GitHub Release rather than in a Marketplace. I am seeing the MarketPlace more like something here to reference the different plugin, panel, dashboards available in different git repositories (GitHub, GitLab, Bitbucket). So the only thing you will need to store in the Marketplace is a sort of document that reference your repository if that makes sense. And then through the MarketPlace, you could be able to download the dashboards. The Marketplace would have an account to be able to get the archive from GitHub

**Reply by Nexucis** [reply]
> What kind of dashboard JSON/YAML format do we persist? 

Probably the one from the Perses API. Then we can provide a transform script to translate it into a CR if you are using the perses-operator

**Reply by bripkens** [reply]
> I believe that even for resources stored into a GitHub (or in GitLab, because why not) repository, you will need to publish a light resource in the Market Place to make the link to that specific repository. And to be able to create this light resource, then a full user/permissions system will be required.

That depends a bit on what is stored within the repo. For example, we can have [meta data files such as this](https://github.com/steadybit/reliability-hub-db/blob/main/actions/com.steadybit.extension_aws.ec2_instance.state/description.yml). That's something I do see as well within the repo. To be concrete:

- `meta.json` (think `package.json`)
- `README.mdx` (long form description for the dashboard)
- `dashboard.json`  ([think this](https://github.com/perses/perses/blob/cfb23a96be3699ca209760f602ac30f6ac297a13/cue/model/api/v1/dashboard_patch.cue#L52-L56))
- `assets/` (preview images)

Personally, being able to leverage GitHub's collaboration model is a strong plus for me. To highlight why I think so, consider [this dashboard in the Grafana dashboard repo](https://grafana.com/grafana/dashboards/1860-node-exporter-full/). How could we mere mortals contribute to that – and how? And what is the license of it? Right now it looks like the nr. 1 Grafana dashboard is owned by a random person on the internet :)

> I am seeing the MarketPlace more like something here to reference the different plugin, panel, dashboards available in different git repositories (GitHub, GitLab, Bitbucket). So the only thing you will need to store in the Marketplace is a sort of document that reference your repository if that makes sense. And then through the MarketPlace, you could be able to download the dashboards. The Marketplace would have an account to be able to get the archive from GitHub

That I think is a good idea. Also to deploy on-premise versions of it with additional proprietary sources.

**Reply by nicolastakashi** [reply]
I think we have an agreement on using Github as the market place file storage.
Every user can host their dashboards on their own repo, but this means user management is on our side.
I think it's also fine I don't think this user management will be very complex.

wdyt?

Btw I love your idea of meta.json @bripkens

**Reply by Nexucis** [reply]
yeah user management is on our side, but I think we can rely on what has been done in Perses.

**Reply by nicolastakashi** [reply]
I need to check, I know we have some OIDC integration.
Are we persisting users?

**Reply by Nexucis** [reply]
yes we are persisting users

**Comment by AntoineThebaud** [comment]
A proposal going more in details, based on @bripkens proposal above:

---

# Description

The marketplace would be a GitHub-based index. This means that a GitHub repo is used as a database of metadata files. The assets to be referenced (namely: dashboards, DaC libraries, and plugin packages) are hosted in the external sources referenced via the metadata files.

1. A new repo `perses/marketplace-index` hosts the list of assets to be presented.
    - Assets are referenced in YAML files, **gathered by owner** - so 1 metadata file per owner. 
    - A user that wants new asset(s) to be referenced has to open a PR to the repo.
    - Example of metadata file:
    ```yaml
    name: "Perses"
    description: "The Perses organization provides a set of official plugins, dashboards and DaC libs to integrate with your Perses instance."
    logo: https://github.com/perses/perses/docs/images/perses.svg

    plugins:
    - name: prometheus
      description: Common plugin module that embeds all the necessary plugins to work with a Prometheus datasource in Perses.
      source:
        github:
          repo: plugins
          filter: prometheus
      tags:
      - tsdb
      - monitoring
    - name: timeserieschart
      description: the standard Time Series chart plugin for Perses
      source:
        github:
          repo: plugins
          filter: timeserieschart
      tags:
      - graph
      - monitoring
    - ...

    dashboards:
    - name: Prometheus Overview
      description: |
        Quick view about Prometheus internal metrics.
        
        ![prometheus_overview](https://my-img-url.com)
        
        The first panel group present an instant view about Prometheus health, the following ones allow you to deep dive in past data with more details
      source: https://github.com/perses/community-dashboards/blob/main/pkg/dashboards/prometheus/prometheus_overview_default.json
      tags:
      - prometheus
      - monitoring

    dashboard_as_code:
    - name: community-dashboards
      description: community-dashboards is designed to provide Prometheus mixins tailored for the Perses platform. Developed with the Perses Go SDK, these dashboards are modular, reusable, and simple to integrate into various observability setups.
      source:
        github:
          repo: community-dashboards
      tags:
      - monitoring
      - tracing
    ```
    - Name of the file = “technical” name of the owner.
    - Content of the file:
      - A first set of attributes about the owner (thus shared by all assets coming after), namely `name`, (optionally) `description` and (optionally) `logo`.
      - Then we have 1 (optional) list for each kind of asset:
        - Common attributes: `name`, `description`, (optional) `tags`.
        - Specific to `plugins`:
          - `source`:
            - only expects a `github` attribute for now. Later other git providers like `gitlab` could be supported.
            - `github` expects 2 attributes, `repo` for the name of the repository hosting this asset & especially its releases, and (optional) `filter` to be used in case of a monorepo hosting several plugins, like [perses/plugins](https://github.com/perses/plugins) is.
        - Specific to `dashboards`:
          - `source` is a URL pointing to the precise JSON file, expected to be of Perses dashboard format.
          - ⚠️ this is unclear if this model would work here! What if you have tons of dashboards?
        - Specific to `dashboard_as_code`:
          - `source` is a URL pointing to the GitHub repo hosting the code for this library. Ideally from its README the user should get access to a proper doc for this lib.
    - Once a PR is opened, a CI runs to check if the information expected to be computed from the modified YAML file(s) are well computed. (see below regarding computed information).

2. A new repo `perses/marketplace` hosts the source code of the marketplace.
    - This application doesn’t need to manage users since publishing assets is done via PRs to `perses/marketplace-index`.
    - For each asset, besides the information coming from the YAML files, the UI would also show computed information. Each info should be cached to avoid spamming the GitHub API.
      - Common information for each asset: 
        - `Last updated`.
          - For plugins & DaC libraries, this would be the date of the latest release available. If no release available, fallback to the date of the last commit.
          - For dashboards, this would be the last modification date of the file.
      - Specific to `Plugins`:
        - A `Versions` section shows up as a table of download links. Each row contains: Version number, a direct download link to the package and a Compatibility column informing about which Perses version this plugin version was made for.
          - The list of versions & direct download links can be retrieved via a proper call to the GitHub API, e.g for TimeSeriesChart https://api.github.com/repos/perses/plugins/releases?q=timeserieschart. In this API call the repo name & the query param corresponds respectively to the `repo` and `filter` attributes defined in the YAML file for this plugin.
          - The Perses-corresponding version (compatibility) is retrieved from the `package.json` of the plugin package (`peerDependencies` field). This is for sure a best-effort  & cautious logic since e.g for MyPlugin v3, even though it depends on Perses v0.52 it could be compatible with previous and/or prior versions of it. We should probably inform clearly the user on the UI about that. 
        - Plugin information like `Plugin kind` (Panel, Datasource, Explore..) or `Display name` are also extracted from `package.json` (of the most recent release available) and shown. In case of a plugin module embedding several plugins (like Prometheus), 1 section appears for each plugin.
      - Specific to `Dashboards`:
        - `Dependencies`, aka which plugins are used/required by this dashboard. This would be computed by parsing the JSON of the dashboard. In terms of the plugins versions to use it’s not something available at the moment but this would be enabled by what’s proposed here https://github.com/perses/perses/discussions/1186#discussioncomment-11972788.
      - Specific to `DaC libraries`:
        - module name. This info would be either extracted either from `go.mod` or `cue.mod`.
        - A `Versions` section shows up as a table. Each row contains the version number and a Compatibility column informing to which Perses version this library version corresponds.
          - The list of versions can be retrieved via a proper call to the GitHub API to get the releases.
          - The Perses-corresponding version (compatibility) is retrieved from the deps of the module file (`go.mod` or `cue.mod`). This is for sure a best-effort & cautious logic since e.g for my-lib v3, even though it depends on Perses v0.52 it could be compatible with previous and/or prior versions of it. We should probably inform clearly the user on the UI about that.
    - The marketplace would come with several capabilities to allow users to search for specific assets. They could filter based on:
      - Name.
      - Type (plugin | dashboard | DaC library).
      - Tags (open question: how to run such search with an SQL-based DB?).
      - Plugin kind (exhaustive enum to this date: `Panel | Datasource | Variable | TimeSeriesQuery | TraceQuery | ProfileQuery | LogQuery | Explore`). We should be careful about the fact that search by plugin kind should also work for multi-plugins module.
      - Owner.
      - Perses version.

# ${\color{green}Pros}$ of this approach
- There is no need for user management or a separate storage mechanism for assets, thus dramatically reducing the scope.
- Clearly shows where a given asset comes from thus allow users to contribute to it eventually - as opposed e.g to the Grafana marketplace where the most well known dashboard (node exporter full) is owned by a random user.

# ${\color{red}Cons}$ of this approach
- May not fly at high scale*:
  - GitHub PR to review for each and every new asset to be published → not suitable with XX PRs per day.
E.g. this approach works well for Prometheus to [reference exporters](https://prometheus.io/docs/instrumenting/exporters/), but of course such strategy wouldn't fly for very dynamic cases like Docker Hub & NPM.
  - Risk of throttling if many (unauthenticated) users download assets from e.g GitHub.
  - *Actually the review could be quite automated here so maybe not a big issue?

# Discarded options

- Having 1 independent metadata file for each asset: we anticipate that it would be more tedious to update for people/orgs providing multiple assets, so 1 file per owner is probably easier to manipulate.

# Out of MVP

Ideas kept for later: 
- Regarding source for plugins: we could support a fallback typeother for any not-supported source. When using other, direct download links would not be generated. other would come with an url field  and it would be up to the user to decide what kind of URL to put there (direct download of latest? Releases page? etc).
- Even with a GitHub-based index like this, we may still want to implement a proper authentication, e.g for the sake of discriminated rate limitation per user or to enable secondary features like mark as favorite, comment, etc.
- On the asset view, show number of views / page hits.
- On the asset view, for dashboards & plugins, show number of downloads.
- On the asset view, for DaC libraries, alternatively to the number of downloads we could retrieve the amount of Dependent packages from https://libraries.io/go.
- The marketplace could be deployed by private companies to reference public assets + their private ones. To allow such setup easily the private marketplace would mirror the assets available from the public instance (= query the same API endpoint than the frontend of the public instance?).

---

Don't hesitate to share your opinion on this, especially about the main remaining point: will it fly to handle dashboards like that?

**Comment by Gladorme** [comment]
Look good to me, maybe add a field "reviewers" with Github pseudos, that will add all "reviewers" as PR reviewer. And if we have a Github bot, maybe we could add a command for reviewers to merge their PR, where they are reviewer, without being collaborator on the repo :)

**Comment by nicolastakashi** [comment]
First, I want to say I really loved the ideas shared earlier. They are clever, well thought out, and were a strong source of inspiration for this proposal.

I'd like to propose something slightly similar in spirit, but diverges in one key aspect:

> Instead of storing artifacts directly in Git or fetching them from source repositories, this design relies on **OCI artifacts hosted on GitHub Container Registry (GHCR)**.

The motivation behind this change is mainly practical:

- Storing or cloning artifacts via Git can become fragile at scale (large files, slow clones, repository drift, rate limits).
- OCI registries are purpose-built for distributing immutable bundles, with built-in versioning, signing, and metadata.
- Using GHCR keeps everything in the GitHub ecosystem while providing stronger security guarantees and simpler automation.

The rest of the model, including open contribution, YAML index entries, CI-based validation, and mirroring, follows the same philosophy discussed before.

## **Overview**
Each resource (dashboard, plugin, or Dashboard-as-Code library) is distributed as an **OCI artifact** stored in GHCR and referenced from the perses/marketplace-index repository.

- **Artifacts** are pushed and signed directly from the contributor’s GitHub repository.
- **Metadata** (owner, type, version, digest, compat, tags) lives in the index repository as a YAML entry.
- **Verification and mirroring** happen automatically when the PR to the index repository is opened.

This keeps ownership decentralized while keeping validation, provenance, and UX centralized.

## **Core Design**

### **1. Single Registry: GHCR**
Only **GitHub Container Registry** is allowed as the publishing backend.

**Why**
- Unified authentication via GitHub OIDC (no manual secrets).
- Automatic provenance: each artifact is traceable to its source repository.
- Simplified verification: one registry API.
- Seamless developer experience: contributors already use GitHub Actions.

Artifacts live under the publisher’s namespace:

```
ghcr.io/<owner>/perses/<kind>/<name>:<version>
```

where:
- **kind** = dash, plugin, or dac
- **version** follows semantic versioning
- **owner** matches the GitHub organization or user
### **2. OCI Artifacts for All Asset Types**
Each resource is published as an **OCI artifact** with a **custom media type** and **annotations**.

|**Kind**|**Media Type**|**Example Payload**|
|---|---|---|
|Dashboard|application/vnd.perses.dashboard.bundle.v1+tar|tar with JSON files, readme, screenshots|
|Plugin|application/vnd.perses.plugin.bundle.v1+tar|compiled plugin bundle|
|DaC Library|application/vnd.perses.dac.bundle.v1+tar|cue/json templates and docs|

**Annotations** (required)
- org.opencontainers.image.title
- org.opencontainers.image.description
- org.opencontainers.image.licenses (SPDX)
- org.opencontainers.image.source (repository URL)
- io.perses.kind = dash | plugin | dac
- io.perses.compat.perses = version constraint
- io.perses.tags = comma-separated tags

### **3. Publisher Flow (via GitHub Action)**
We will provide an official composite action: **perses-actions/publish-and-propose**
It performs all steps end to end:
1. Builds the bundle (dashboard JSON, plugin zip, etc.).
2. Pushes to GHCR with the correct artifactType and annotations.
3. Signs the digest with **cosign keyless** (OIDC).
4. Optionally attaches SBOM and SLSA provenance as OCI referrers.
5. Generates the index entry YAML.
6. Opens a PR to perses/marketplace-index automatically.

**Example usage**

```
- uses: perses-actions/publish-and-propose@v1
  with:
    kind: dash
    name: acme-app-dashboard
    version: 1.2.3
    bundle: out/acme-app-dashboard.tgz
    title: "Acme App Dashboard"
    description: "Overview of latency, errors, and saturation."
    license: Apache-2.0
    compat_perses: ">=0.46 <0.50"
    tags: "kubernetes,latency,p99"
```

Contributors get a single declarative step with no manual oras, cosign, or PR creation.
### **4. Index Repository and Validation**

Each submission opens a PR to perses/marketplace-index containing a YAML entry:

```
owner: acme
name: acme-app-dashboard
type: dash
version: 1.2.3
source_repo: https://github.com/acme/observability
registry_ref: ghcr.io/acme/perses/dash/acme-app-dashboard@sha256:<digest>
license: Apache-2.0
compat:
  perses: ">=0.46 <0.50"
tags: ["kubernetes","latency","p99"]
```

The **index CI pipeline** then:

1. Pulls the referenced digest from GHCR.
2. Verifies the cosign signature (issuer = GitHub, subject = source repository).
3. Validates media type and required annotations.
4. Runs linting and sanitization (schema, license, size, allowed files).
5. Optionally verifies SLSA provenance and SBOM for plugins.
6. Mirrors the artifact under Perses’s cache registry:

```
ghcr.io/perses/cache/<owner>/<kind>/<name>@sha256:<digest>
```
7. Merges when all checks pass.

### **5. Consumer Behavior (UI and CLI)**
- Always install by **digest**, not tag.
- Prefer mirror_ref (Perses cache) for reliability.
- Verify cosign signature at install time.
- Use annotations to display title, compatibility, license, and tags.
## **✅ Benefits**

|**Category**|**Value**|
|---|---|
|**Security**|Strong provenance via keyless signatures, optional SLSA, and digest pinning|
|**Simplicity**|One registry, one GitHub Action, one PR flow|
|**Scalability**|No manual review needed beyond automated CI checks|
|**Ownership**|Publishers own their artifacts under their own organizations|
|**Reliability**|Perses mirrors artifacts automatically|
|**Consistency**|All asset types use OCI with a shared metadata schema|

## **⚠️ Limitations and Considerations**

- Requires GHCR support for OCI referrers (already available).
- Marketplace maintainers must manage cache retention policies.
- Documentation must clearly specify required annotations.
- Publishers need packages:write permission in their GitHub organization.

## **🧩 Future Extensions**

- Automatic artifact scanning and linting during index CI.
- A “trusted publisher” mode for auto-merge when all checks pass.
- Marketplace UI enriched with stars, downloads, and popularity metrics.
- Optional on-prem mirroring through OCI sync.
- Publish rules (only push on tags, require immutable tags)


Once we get an agreement we should start writing a design document

**Reply by AntoineThebaud** [reply]
Coming back to this quickly, I'd say that OCI artifacts could fit well for the dashboards & especially plugins, but not for Dashboard-as-Code libraries: in this case the marketplace is just there to be an index, the users would then rely on the libs referenced there by simply leveraging the dependency management of the language (whether it's Go or CUE).
